### DISCUSSION TITLE
request becomes "https" unintentionally when accessing from mobile

### DISCUSSION THREAD

**Discussion by KeisukeNagakawa** [vercel/next.js#36875] [Help] [answered]
When I am developping nextjs project in local environment on my Mac, I tried to access the server(= `http://192.168.1.60:3000`) from my iPhone which is connected with the same wifi from my Mac but it failed with the following error:

```
FetchError: request to https://192.168.1.60:3000/api/issues/list failed, reason: write EPROTO 4419700096:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:332
```

It is strange that the request endpoint is becoming `https`, not `http`. How can I fix this error and make it possible to access localhost server from my mobile device? 

Thanks!

**Comment by icyJoseph** [comment, marked as answer]
Hi,

I have an iPhone and a Macbook and I cannot reproduce this. Tested both in Chrome and Safari in iOS. It's the first time I see an issue like this, I usually have an Android and iPhone, side by side connected to my local server, and it just works. I don't have any special settings on either phone though (privacy, and other stuff, is set to defaults).

Through middleware I have confirmed that the request is still `http`.

My guess is that there's a quite a bit going on your app. Do you think I could take a look? Is there any reproduction repository?

Possible factors:

- Some redirect is upgrading the request
- You may have an iphone privacy setting that upgrades requests
- Perhaps you have more moving parts? Caddy, docker, etc?
- You have set SSL yourself?


## Resolution

To make requests to `api/routes` from `GSSP`, this line was used, which when accessed from another device, via IP, makes `isLocalhost` false. If `isLocalhost` is false, the `https` protocol was used when creating an absolute url to reach `api/routes`.

```ts
  const isLocalhost = host.indexOf("localhost") === 0
```

> In this case host is 192.168.0.22:3000, you know, the IP you input into the iPhone.

So then it is not localhost, and it uses https, there you go :) - just try to fix this, but overall, trying to reach for api/routes from GSSP is not a good idea. Just connect to the service directly.

**Comment by KeisukeNagakawa** [comment]
hum...thanks.  Upon further investingation, I found that I could access nextjs page which does not call api.

1.Some redirect is upgrading the request 
-> No

2. You may have an iphone privacy setting that upgrades requests
-> might be...but I can not confirm it. URL is https.

3. Perhaps you have more moving parts? Caddy, docker, etc?
-> No. I am running it on mac with vs code without any container service like Docker.

4. You have set SSL yourself?
-> No.

**Reply by icyJoseph** [reply]
> 2. You may have an iphone privacy setting that upgrades requests
>    -> might be...but I can not confirm it. URL is https.

What do you mean? "URL is https"? 

How are you making the request to `/api/issues/list`?

**Reply by KeisukeNagakawa** [reply]
>What do you mean? "URL is https"?
My mistake. HTTP. I am sorry.

>How are you making the request to /api/issues/list?
Like this
```
import { withAuthUserTokenSSR } from "next-firebase-auth"
import { FC } from "react"
import getAbsoluteURL from "utils/getAbsoluteURL"

const Page: FC = () => {
  return <div>test</div>
}

export const getServerSideProps = withAuthUserTokenSSR({})(
  async ({ AuthUser, req, resolvedUrl }) => {
    // list and return
    const path = getAbsoluteURL(`/api/issues/list`, req)
    const res = await fetch(path)
    console.log({ res })

    // ...
    return { props: { a: "a" } }
  }
)

export default Page

```

thanks!

**Reply by icyJoseph** [reply]
Aha, here's the issue. Making requests to `api` routes from `GSSP` is not something you should do. It'll be buggy.

You are supposed to connect directly to the service under `/api/issues/list`, and get the data from there.

I'll update with more, shortly.

### Edit

The pattern to be enforced is something like I document here: https://github.com/vercel/next.js/discussions/35532#discussioncomment-2420969, with CodeWars data from an `api/route` and from `GSSP`.

I'd like to see the `getAbsoluteURL` implementation, to know what set you up for trouble.

**Reply by KeisukeNagakawa** [reply]
Wow!!!! Then I have to change whole my code....hahaha!! 

> `getAbsoluteURL`

```
/* globals window */

const getAbsoluteURL = (url: string, req?: any) => {
  let host
  if (req) {
    host = req.headers.host
  } else {
    if (typeof window === "undefined") {
      throw new Error(
        'The "req" parameter must be provided if on the server side.'
      )
    }
    host = window.location.host
  }
  const isLocalhost = host.indexOf("localhost") === 0
  const protocol = isLocalhost ? "http" : "https"
  return `${protocol}://${host}${url}`
}

export default getAbsoluteURL

```

**Reply by icyJoseph** [reply]
```js
  const isLocalhost = host.indexOf("localhost") === 0
```

In this case `host` is `192.168.0.22:3000`, you know, the IP you input into the iPhone.

So then it is not localhost, and it uses `https`, there you go :) - just try to fix this, but overall, trying to reach for `api/routes` from GSSP is not a good idea. Just connect to the service directly. 

If you are worried about code leaking to the client, take a look at https://next-code-elimination.vercel.app/ 

In production, your server might be behind a reverse proxy and bunch of other configs, so there's no guarantee that the host value will be what the user sees or what you think it'll be. Sometimes it'll be, sometimes it won't...

**Reply by KeisukeNagakawa** [reply]
@icyJoseph I am really sorry but inside getAbsoluteURL, I change http into https if it is not localhost... so this was the cause! 

But, as you have said, I think I have to re-write code so as not to call api from `GSSP`. this is going to be a tough work for me....

**Reply by icyJoseph** [reply]
hehe my eyes went directly to that line, see my comment above :)

**Reply by KeisukeNagakawa** [reply]
Oh I found same problem on same time. Thank you @icyJoseph  anyway. You gave me lots of hints, I appreciate it.

**Reply by icyJoseph** [reply]
Refactoring this things often makes it nicer in the long run :) - Also remember, that is also impossible to call `api/routes` from GetStaticProps, because the server is not even running, in case you ever want to do that as well.

**Reply by icyJoseph** [reply]
Don't forget to mark the discussion as answered. Have a good day!

**Comment by KeisukeNagakawa** [comment]
Error shown in my iPhone is this.
```
 1 of 1 unhandled error
Server Error
FetchError: request to https://192.168.1.70:3000/api/issues/list failed, reason: write EPROTO 4419634560:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:332:
This error happened while generating the page. Any console logs will be displayed in the terminal window.
Call Stack
ClientRequest.<anonymous>

file:///Users/kay/Code/aaa/node_modules/next/node_modules/node-fetch/lib/index.js (1461:11)
ClientRequest.emit

node:events (390:28)
ClientRequest.emit

node:domain (475:12)
TLSSocket.socketErrorListener

node:_http_client (447:9)
TLSSocket.emit

node:events (390:28)
TLSSocket.emit

node:domain (475:12)
emitErrorNT

node:internal/streams/destroy (157:8)
emitErrorCloseNT

node:internal/streams/destroy (122:3)
processTicksAndRejections

node:internal/process/task_queues (83:21)
```
